sql/moz-fx-data-shared-prod/subscription_platform/nonprod_stripe_subscriptions_history/view.sql

```diff
@@ -0,0 +1,294 @@
+CREATE OR REPLACE VIEW
+  `moz-fx-data-shared-prod.subscription_platform.nonprod_stripe_subscriptions_history`
+AS
+WITH subscriptions_history AS (
+  SELECT
+    customer_id,
+    id AS subscription_id,
+    _fivetran_synced AS synced_at,
+    _fivetran_start AS valid_from,
+    LEAD(_fivetran_start) OVER (PARTITION BY id ORDER BY _fivetran_start) AS valid_to,
+    created,
+    trial_start,
+    trial_end,
+    COALESCE(trial_end, start_date) AS subscription_start_date,
+    cancel_at,
+    cancel_at_period_end,
+    canceled_at,
+    JSON_VALUE(metadata, "$.cancelled_for_customer_at") AS canceled_for_customer_at,
+    ended_at,
+    status,
+    TIMESTAMP_SECONDS(
+      CAST(JSON_VALUE(metadata, "$.plan_change_date") AS INT64)
+    ) AS plan_change_date,
+    JSON_VALUE(metadata, "$.previous_plan_id") AS previous_plan_id,
+  FROM
+    `dev-fivetran`.stripe_nonprod.subscription_history
+),
+subscriptions_history_with_lead_plan_metadata AS (
+  SELECT
+    *,
+    LEAD(plan_change_date) OVER (
+      PARTITION BY
+        subscription_id
+      ORDER BY
+        valid_from
+    ) AS lead_plan_change_date,
+    LEAD(previous_plan_id) OVER (
+      PARTITION BY
+        subscription_id
+      ORDER BY
+        valid_from
+    ) AS lead_previous_plan_id,
+  FROM
+    subscriptions_history
+),
+subscriptions_history_with_plan_ends AS (
+  SELECT
+    *,
+    -- A new `previous_plan_id` value means the previous row was the last row with that plan.
+    IF(
+      lead_previous_plan_id IS DISTINCT FROM previous_plan_id,
+      STRUCT(lead_previous_plan_id AS plan_id, lead_plan_change_date AS plan_ended_at),
+      (NULL, NULL)
+    ).*
+  FROM
+    subscriptions_history_with_lead_plan_metadata
+),
+subscriptions_history_with_previous_plan_ids AS (
+  -- Fill in previous `plan_id` values by looking forward to the next non-null `plan_id`.
+  SELECT
+    * EXCEPT (plan_id),
+    FIRST_VALUE(plan_id IGNORE NULLS) OVER (
+      PARTITION BY
+        subscription_id
+      ORDER BY
+        valid_from
+      ROWS BETWEEN
+        CURRENT ROW
+        AND UNBOUNDED FOLLOWING
+    ) AS plan_id,
+    LAST_VALUE(plan_ended_at IGNORE NULLS) OVER (
+      PARTITION BY
+        subscription_id
+      ORDER BY
+        valid_from
+      ROWS BETWEEN
+        UNBOUNDED PRECEDING
+        AND 1 PRECEDING
+    ) AS plan_started_at,
+  FROM
+    subscriptions_history_with_plan_ends
+),
+subscription_items AS (
+  SELECT
+    id AS subscription_item_id,
+    subscription_id,
+    plan_id,
+  FROM
+    `dev-fivetran`.stripe_nonprod.subscription_item
+  -- ZetaSQL requires QUALIFY to be used in conjunction with WHERE, GROUP BY, or HAVING.
+  WHERE
+    TRUE
+  QUALIFY
+    -- With how our subscription platform currently works each Stripe subscription should
+    -- only have one subscription item, and we enforce that so the ETL can rely on it.
+    1 = COUNT(*) OVER (PARTITION BY subscription_id)
```

Comment:
Rather than throwing out rows that violate our condition I think we should throw an error, and thus get notified of ETL failing because this assumption is incorrect.
```suggestion
    IF(COUNT(*) OVER (PARTITION BY subscription_id) > 1, ERROR("stripe subscription contains multiple items"), TRUE)
```

Reply:
Unfortunately that would already fail with the error because a test subscription with multiple subscription items exists in the non-prod environment.  And while we could currently add the error logic in the production view, I'm a little leery of some future test/experimental/accidental Stripe setup breaking the ETL, forcing us to immediately scramble to address it.

Other ideas would be to create a "check" ETL that would run independently and error under the same conditions, or try using something like Great Expectations or the [data integrity monitoring PoC](https://github.com/mozilla/dim/).  Thoughts?

Reply:
no change for now is fine then.